### ISSUE TITLE
Use previously fetched Google Oauth Token

### ISSUE THREAD

**Issue by Jeremlicious** [kreait/firebase-php#404]
Hi, we use kreait/firebase-php 4.43 in SQS consumer and would like to save some queries to Google.

When doing this :
```php
$firebaseClient->getDatabase()->getReference('path1')->getValue();
$firebaseClient->getDatabase()->getReference('path1')->getValue();
```

the method `Google\Auth\OAuth2::fetchAuthToken` is called twice and two HTTP auth queries are sent to Google.
The received tokens have **expires_in=3599**.

How to use the first token for the second query to Google ?

**Comment by Jeremlicious** [comment]
Hi, thanks ! Yes I will be able to test.
Can't upgrade to 5.0 yet, we are still using php 7.1.19

**Comment by Jeremlicious** [comment]
Hi, thanks !
Here is the tests I made :

```php
$database->getReference('path1')->getValue();
sleep(30);
$database->getReference('path2')->getValue();
```
I get only one call to `Google\Auth\Oauth2::fetchAuthToken`
According to `withEnabledDebug` the same Oauth token is sent ✔️ 

```bash
*   Trying 35.201.97.x...
* Connected to myproject.firebaseio.com (35.201.97.x) port 443 (#0)
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use http/1.1
* Server certificate:
*        subject: C=US; ST=California; L=Mountain View; O=Google LLC; CN=firebaseio.com
*        start date: Feb 20 19:36:06 2020 GMT
*        expire date: Feb 18 19:36:06 2021 GMT
*        subjectAltName: testjerem-92843.firebaseio.com matched
*        issuer: C=US; O=Google Trust Services; CN=GTS CA 1O1
*        SSL certificate verify ok.
> GET /path1.json HTTP/1.1
Host: myproject.firebaseio.com
User-Agent: GuzzleHttp/6.5.1 curl/7.47.0 PHP/7.1.30-1+ubuntu16.04.1+deb.sury.org+1
authorization: Bearer xxxxxxx.SAME.TOKEN.xxxxxxx

< HTTP/1.1 200 OK
< Server: nginx
< Date: Sat, 04 Apr 2020 22:36:14 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 4
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Cache-Control: no-cache
< Strict-Transport-Security: max-age=31556926; includeSubDomains; preload
<
* Connection #0 to host myproject.firebaseio.com left intact
```

```bash
* Found bundle for host myproject.firebaseio.com: 0x556d25297180 [can pipeline]
* Re-using existing connection! (#0) with host myproject.firebaseio.com
* Connected to myproject.firebaseio.com (35.201.97.x) port 443 (#0)
> GET /path2.json HTTP/1.1
Host: myproject.firebaseio.com
User-Agent: GuzzleHttp/6.5.1 curl/7.47.0 PHP/7.1.30-1+ubuntu16.04.1+deb.sury.org+1
authorization: Bearer xxxxxxx.SAME.TOKEN.xxxxxxx

< HTTP/1.1 200 OK
< Server: nginx
< Date: Sat, 04 Apr 2020 22:36:44 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 4
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Cache-Control: no-cache
< Strict-Transport-Security: max-age=31556926; includeSubDomains; preload
<
* Connection #0 to host myproject.firebaseio.com left intact
```

It seems **ok** to me  please note that I haven't tested more for regressions.

By the way, it will be usefull to enable Guzzle request debug onGoogle Oauth call. We sometime get weird connectvity and SSL issues from AWS.
I could not figure out how to enable this debug from firebase-php.

**Comment by jeromegamez** [comment]
Thanks for testing! If "ok" means "good", I'll take it 😅

Could you please `composer update` once more? The calls to fetch the OAuth tokens should now respect the `$factory->withEnabledDebug()` option as well. (I tested it already, but would like you confirm before I put on the finishing touches 😅)

**Comment by Jeremlicious** [comment]
Hi, yeah it means good !
Just that I did not make extensive testing. For example what happens when trying to get another kind of Google Oauth token if there is. Or some different scope.
Great to have `withEnableDebug()` on the Oauth token fetch! That will help next time we get connectivity issues.

By the way, does this update means that
```php
$factory->withHttpClientConfig([RequestOptions::TIMEOUT => 4]);
```
will be passed on to Google Api client ?
I saw that `withHttpClientConfig` is deprecated. What is the defaut timeout then ? Is it possible to change it ?

**Comment by jeromegamez** [comment]
I deprecated `withHttpClientConfig()` and `withHttpClientMiddlewares()` because I created them before having added the Firestore and Cloud Storage components (which rely on the official google libraries).

The Google libraries don't allow customizations like this, so, unfortunately, `withEnabledDebug()` and `withHttpProxy()` will continue to have no effect there, but the expectations are perhaps not as "big" as with the other methods promising much more 🙈

Also, it was possible to break the SDK by adding config options and middleware - in the past, I tried to help developers debug a problem, and it turned out that they had set the timeouts too low or added a middleware that broke the requests. So it's also a matter of keeping the SDK stable/preventing situations where it becomes less stable 😅.

To answer your question directly: `withHttpClientConfig()`, `withHttpClientMiddlewares()`, `withEnabledDebug()` and `withHttpProxy()` will have an effect on all requests except to Firestore and Cloud Storage.

But I will be looking for ways to re-add customization options - I have many ideas, but they will require step by step changes, I'm not ready for a full rewrite yet 😄

**Comment by jeromegamez** [comment]
I merged the changes and released them as 4.44.0 and 5.1.0, so I'm closing this ticket as the original issue is resolved, but please feel free to add another comment if you'd like to add something!

Thank you very much for testing and the cooperation in general! 🌺

**Comment by Jeremlicious** [comment]
Hi, we just released 4.44 in production. Works like a charm. Thanks again !
